lib/ruby_terraform/commands/base.rb

```diff
@@ -39,6 +39,8 @@ def execute(opts = {})
           stderr: stderr
         )
         do_after(opts)
+      rescue Open4::SpawnError
+        @logger.error "Command '#{command.to_s}' has failed."
```

Comment:
This will potentially expose sensitive information into whatever is running terraform. I think having this error handling is really useful but can't think of a good way to prevent the credential leakage. Maybe leaving the log statement at debug for now is the best we can do until we come up with a better approach.

In the longer term, maybe there's something about marking certain vars as sensitive so that they can be filtered out of any logging.

Reply:
I also wonder if the exception should be rethrown since the execution should exit with a non-zero exit status.

Reply:
> This will potentially expose sensitive information into whatever is running terraform. I think having this error handling is really useful but can't think of a good way to prevent the credential leakage. Maybe leaving the log statement at debug for now is the best we can do until we come up with a better approach.
>
Yes, I think one ood way is keep `error` call and Just inform about what command went wrong. Something like
`Terraform failed while running *apply*` 
or similar message and do not show the whole command. 
> In the longer term, maybe there's something about marking certain vars as sensitive so that they can be filtered out of any logging.
>
That would be handy, not sure terraform people even consider that kind of sensitive issues

Reply:
> Yes, I think one good way is keep error call and report what command went wrong. Something like 'Terraform failed while running *apply*' or similar message and do not show the whole command.

Yes, this approach would work too. It's not very easy to get hold of the terraform command from `RubyTerraform::Commands::Base` though. You might need to add a protected `command_name` method on `Base` and implement it on each command. I'm also unsure how to handle this in the nested command case. For example the workspace command isn't currently broken out for each sub-sub-command.

What are your thoughts on how to exit with a non-zero exit code when an exception is thrown. This would be a breaking change and my deployment pipelines rely on the command failing to prevent further execution. My feeling is `RubyTerraform` should continue to throw and consuming code can choose to handle that exception if needed. Alternatively the exception could be mapped to a documented `RubyTerraform` exception to make it easier for consumers to know what to catch.

> That would be handy, not sure terraform people even consider that kind of sensitive issues

Terraform currently provides this on outputs rather than variables. See [the output docs](https://www.terraform.io/docs/configuration/outputs.html#sensitive-suppressing-values-in-cli-output) for details. This is quite important for me since I need everyone to have access to the production deployment job log output but few people to have access to the production credentials.

Reply:
> > Yes, I think one good way is keep error call and report what command went wrong. Something like 'Terraform failed while running _apply_' or similar message and do not show the whole command.
> 
> Yes, this approach would work too. It's not very easy to get hold of the terraform command from `RubyTerraform::Commands::Base` though. 
>
As an attempt to check, I got it from `command.to_s` (haven't tested for edge cases if any but for normal case, it works)  
>You might need to add a protected `command_name` method on `Base` and implement it on each command. I'm also unsure how to handle this in the nested command case. For example the workspace command isn't currently broken out for each sub-sub-command.
> 
I like that idea.
> What are your thoughts on how to exit with a non-zero exit code when an exception is thrown. This would be a breaking change and my deployment pipelines rely on the command failing to prevent further execution. My feeling is `RubyTerraform` should continue to throw and consuming code can choose to handle that exception if needed. Alternatively the exception could be mapped to a documented `RubyTerraform` exception to make it easier for consumers to know what to catch.
> 
I like the documented exception and consumers decide.
> > That would be handy, not sure terraform people even consider that kind of sensitive issues
> 
> Terraform currently provides this on outputs rather than variables. See [the output docs](https://www.terraform.io/docs/configuration/outputs.html#sensitive-suppressing-values-in-cli-output) for details. This is quite important for me since I need everyone to have access to the production deployment job log output but few people to have access to the production credentials.
>
That looks good but it seems probably not enough ? It does what we want in `apply` but there will be other commands potentially and that value can still be shown in the CLI.

Reply:
> As an attempt to check, I got it from command.to_s (haven't tested for edge cases if any but for normal case, it works)

`command.to_s` will return the full command including all options, thus including the sensitive information. Really we just want the terraform subcommand so something like `apply` or `workspaces list`.

> > > That would be handy, not sure terraform people even consider that kind of sensitive issues
> >
> > Terraform currently provides this on outputs rather than variables. See the output docs for details. This is quite important for me since I need everyone to have access to the production deployment job log output but few people to have access to the production credentials.
>
> That looks good but it seems probably not enough ? It does what we want in apply but there will be other commands potentially and that value can still be shown in the CLI.

Yes, what I meant to say was that Terraform does take some measures to protect sensitive values. Terraform also provides the var files mechanism which allows you to keep sensitive vars off the command line. You are correct that we would need to implement something in addition to what terraform provides if we wanted similar protections.

Reply:
> > As an attempt to check, I got it from command.to_s (haven't tested for edge cases if any but for normal case, it works)
> 
> `command.to_s` will return the full command including all options, thus including the sensitive information. Really we just want the terraform subcommand so something like `apply` or `workspaces list`.
> 
Yes, I didn't do it like that (that is how currently shows in this PR) but finding the very next word after `terraform` in that string. As I said, I have not tested fully.
I like the idea of the `command_name` method, will take a look into that.

> > > > That would be handy, not sure terraform people even consider that kind of sensitive issues
> > > 
> > > 
> > > Terraform currently provides this on outputs rather than variables. See the output docs for details. This is quite important for me since I need everyone to have access to the production deployment job log output but few people to have access to the production credentials.
> > 
> > 
> > That looks good but it seems probably not enough ? It does what we want in apply but there will be other commands potentially and that value can still be shown in the CLI.
> 
> Yes, what I meant to say was that Terraform does take some measures to protect sensitive values. Terraform also provides the var files mechanism which allows you to keep sensitive vars off the command line. You are correct that we would need to implement something in addition to what terraform provides if we wanted similar protections.
>
That would be nice as then `RubyTerraform` wouldn't have to care about that, meaning the logs or any other case where info is shown. 
That can be a different PR, IMO.

Reply:
> That can be a different PR, IMO.

Yes, agree.